indigo_optional_drivers/ccd_rpi/indigo_ccd_rpi.cpp

```diff
@@ -0,0 +1,1799 @@
+
+            pixelFormats.push_back(format);
+        }
+    }
+    return pixelFormats;         
+}
+bool RPiCamera::HasAutoFocus() const
+ {
+    if (!m_cameraAcquired) {
+        INDIGO_DRIVER_ERROR(DRIVER_NAME, "Camera not acquired" );
+        return false;
+    }
+    // Check if the camera has a focus control
+    const ControlInfoMap &controls = m_camera->controls();
+    return controls.find(&controls::AfMode) != controls.end();
+ }
+ bool RPiCamera::ReadFrame(FrameData *frameData){
+    std::lock_guard<std::mutex> lock(m_freeRequestsMutex);
+    // int w, h, stride;
+    if (!m_requestQueue.empty()){
+        Request *request = this->m_requestQueue.front();
+
+        const Request::BufferMap &buffers = request->buffers();
+        for (auto it = buffers.begin(); it != buffers.end(); ++it) {
+            FrameBuffer *buffer = it->second;
+            for (unsigned int i = 0; i < buffer->planes().size(); ++i) {
+                const FrameBuffer::Plane &plane = buffer->planes()[i];
+                const FrameMetadata::Plane &meta = buffer->metadata().planes()[i];
+                
+                void *data = m_mappedBuffers[plane.fd.get()].first;
+                int length = std::min(meta.bytesused, plane.length);
+
+                frameData->size = length;
+                frameData->imageData = (uint8_t *)data;
+            }
+        }
+        this->m_requestQueue.pop();
+        frameData->request = (uint64_t)request;
+        return true;
+    } else {
+        Request *request = nullptr;
+        frameData->request = (uint64_t)request;
+        return false;
+    }
+}
+Stream *RPiCamera::VideoStream(uint32_t *w, uint32_t *h, uint32_t *stride) const
+{
+	StreamDimensions(m_stream, w, h, stride);
+	return m_stream;
+}
+Size RPiCamera::GetMaxImageSize() const
+{
+    if (!m_cameraAcquired) {
+        INDIGO_DRIVER_ERROR(DRIVER_NAME, "Camera not acquired" );
+        return {};
+    }
+
+    std::unique_ptr<CameraConfiguration> config = m_camera->generateConfiguration({ StreamRole::StillCapture });
+    libcamera::CameraConfiguration::Status status = config->validate();
+    if (status != libcamera::CameraConfiguration::Valid) {
+        INDIGO_DRIVER_ERROR(DRIVER_NAME, "Camera configuration is not valid" );
+        return {};
+    }
+    StreamConfiguration &streamConfig = config->at(0);
+    // Get the format
+    PixelFormat pixelFormat = streamConfig.pixelFormat;    
+    // Access the format map to find all supported sizes
+    const StreamFormats &formats = config->at(0).formats();
+    // Get all sizes for the pixel format
+    std::vector<Size> sizes = formats.sizes(pixelFormat); 
+    // Find the largest one
+    auto maxSize = *std::max_element(sizes.begin(), sizes.end(), [](const Size &a, const Size &b) {
+        return a.width * a.height < b.width * b.height;
+    });
+
+    config.reset(); // Clean up the configuration
+
+    return maxSize;
+
+}
+bool RPiCamera::af()
+{
+    if(HasAutoFocus())
+    {
```

Comment:
please use  
bool RPiCamera::af() {
    if (HasAutoFocus()) {

note the space between if and ()